dagger/storageproofs/bls.nim

```diff
@@ -0,0 +1,257 @@
+## Nim-POS
+## Copyright (c) 2021 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+# Implementation of the BLS-based public PoS scheme from 
+# Shacham H., Waters B., "Compact Proofs of Retrievability" 
+# using pairing over BLS12-381 ECC 
+# 
+# Notation from the paper
+# In Z:
+# - n: number of blocks
+# - s: number of sectors per block
+#
+# In Z_p:
+# - m_{ij}: sectors of file
+# - α: pos secret key
+# - name
+# - μ_j 
+#
+# In G_1: multiplicative cyclic group
+# - H: {0,1}∗ →G_1 
+# - u_1,…,u_s ←R G_1 
+# - σ_i: authenticators
+# - σ: part of proof
+#
+# In G_2: multiplicative cyclic group
+# - g: generator
+# - v ← g^α: pos public key
+#
+# In G_T:
+# - two pairings for validation
+
+import blscurve
+import blscurve/blst/blst_abi
+import random
+
+const sectorsperblock = 1024.int64
+const bytespersector = 31 # r is 255 bits long 
+const querylen = 22
+
+type ZChar = array[bytespersector, byte]
+
+type SecretKey = object
+  signkey: blscurve.SecretKey
+  key: blst_scalar
+
+type PublicKey = object
+  signkey: blscurve.PublicKey
+  key: blst_p2
+
+type TauZero = object
+  name: array[512,byte]
+  n:    int64
+  u:    seq[blst_p1]
+
+type Tau = object
+  t: TauZero
+  signature: array[512, byte]
+
+proc fromBytesBE(a: array[32, byte]): blst_scalar =
+  blst_scalar_from_bendian(result, a)
+  doAssert(blst_scalar_fr_check(result).bool)
+
+proc fromBytesBE(a: openArray[byte]): blst_scalar =
+  var b: array[32, byte]
+  doAssert(a.len <= b.len)
+  let d = b.len - a.len
+  for i in 0 ..< a.len:
+    b[i+d] = a[i]
+  blst_scalar_from_bendian(result, b)
+  doAssert(blst_scalar_fr_check(result).bool)
+
+proc getSector(f: File, blockid: int64, sectorid: int64, spb: int64): ZChar =
+  f.setFilePos(blockid * spb + sectorid)
+  let r = f.readBytes(result, 0, sizeof(result))
+
+proc rndScalar(): blst_scalar =
+  var scal{.noInit.}: array[32, byte]
+  var scalar{.noInit.}: blst_scalar
+
+  while true:
+    for val in scal.mitems:
+      val = byte rand(0xFF)
+    scalar.blst_scalar_from_bendian(scal)
+    if blst_scalar_fr_check(scalar).bool:
+      break
+
+  return scalar
+
+proc rndP2(): (blst_p2, blst_scalar) =
+  var x{.noInit.}: blst_p2
+  x.blst_p2_from_affine(BLS12_381_G2) # init from generator
+  let scalar = rndScalar()
+  x.blst_p2_mult(x, scalar, 255) 
+  return (x, scalar)
+
+proc rndP1(): (blst_p1, blst_scalar) =
+  var x{.noInit.}: blst_p1
+  x.blst_p1_from_affine(BLS12_381_G1) # init from generator
+  let scalar = rndScalar()
+  x.blst_p1_mult(x, scalar, 255) 
+  return (x, scalar)
+
+proc posKeygen(): (blst_p2, blst_scalar) =
+  rndP2()
+
+proc keygen*(): (PublicKey, SecretKey) =
+  var pk: PublicKey
+  var sk: SecretKey
+  var ikm: array[32, byte]
+  var RNG = initRand(0xFACADE)
```

Comment:
Just a heads up (and it's probably fine for a first pass PoC), but the std random functions aren't cryptographically secure - we're using `bearssl` as our crypto secure random generator instead. There is `dagger/rng.nim`, which exposes a couple of convenience functions around the bearssl HMAC primitive, preferably, we should be using that and there should be a few example uses in this codebase already.

Saddly, we haven't standardized it across all ouf our projects yet, so you'll fine similar reimplementations across other projects.

Reply:
Thanks @dryajov. I've used random only for initial testing of the calculation. I will change all random functions in a subsequent patch to crypto secure ones.

Reply:
Cool, yep I figured it was just a first pass, just noting here for clarity 😉